---
title: "Figure 5"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---

####Enhancer TF factor analysis
Here's the idea is to find significantly enriched TFs in gained or lost enhancers
For that we need gained and lost enhancers by random forest classifier 
and TF binding sites from UCSC table browser

load libraries
```{r}
library(ggplot2)
library(dplyr)
library(zoo)
library(stringi)
library(stringr)
library(data.table)
library(GenomicRanges)
library(tidyr)
library(reshape)
```


#selecting gain+loss as bg
```{r}

#load TF filenormalized to 400 bp from middle
##hg19 based
TF_norm <- fread("TF.binding.sites.extended.range.200bp.from.middle.bed")

colnames(TF_norm) <- c("chr","start","end","TF_name","score")

TF_norm <- makeGRangesFromDataFrame(TF_norm, keep.extra.columns=TRUE)

predicted_element <- fread("enhancers/02.enhancer_vs_expression.table.csv")

predicted_element <- unique(predicted_element)
element_width=1000



predicted_element$start <- predicted_element$start+ as.integer((predicted_element$end-predicted_element$start)/2) 
predicted_element$end <- predicted_element$start 
predicted_element$start <- predicted_element$start - (element_width/2) +1
predicted_element$end  <- predicted_element$end + (element_width/2)


predicted_element<- dplyr::select(predicted_element, c("chr", "start", "end", "Peak","predicted_gain","predicted_loss"))
predicted_element <- subset(predicted_element, Peak!="")

gain <- subset(predicted_element, predicted_gain==T & predicted_loss==F)

lost <- subset(predicted_element, predicted_gain==F & predicted_loss==T)




gain <- makeGRangesFromDataFrame(gain, keep.extra.columns=TRUE)
lost <- makeGRangesFromDataFrame(lost, keep.extra.columns=TRUE)
	
TFList <- TF_norm %>% as.data.frame() %>%dplyr::select(TF_name) %>% unique()
TFList2 <- TF_norm %>% as.data.frame() %>%dplyr::select(TF_name) %>% unique()
#i=1
for(i in 1:dim(TFList)[1]){
  TF_overlap <- subset(TF_norm, TF_name==TFList[i,1])
  overlap_TF_gain <- countOverlaps(  gain, TF_overlap, ignore.strand=TRUE,type="any")
    overlap_TF_gain <- as.data.frame(overlap_TF_gain)
 overlap_TF_gain <- subset(overlap_TF_gain, overlap_TF_gain > 0) %>% nrow()
 
overlap_TF_loss <- countOverlaps(lost , TF_overlap,  ignore.strand=TRUE,type="any")
  overlap_TF_loss <- as.data.frame(overlap_TF_loss)
 overlap_TF_loss <- subset(overlap_TF_loss, overlap_TF_loss > 0) %>% nrow() 
 
 TFList[i,2] <- overlap_TF_gain
 TFList[i,3] <- overlap_TF_loss
 

 
 TFList[i,4]<-phyper((overlap_TF_gain)-1,length(gain),length(lost),overlap_TF_gain+overlap_TF_loss, lower.tail = FALSE)
 
 TFList[i,5]<-phyper((overlap_TF_loss)-1,length(lost),length(gain),overlap_TF_gain+overlap_TF_loss, lower.tail = FALSE)
  V1 = overlap_TF_gain
  V2 = length(gain)-overlap_TF_gain
  V3 = overlap_TF_loss
  V4 = length(lost)-overlap_TF_loss
  
  TFList2[i,2] <- V1
 TFList2[i,3] <- V2
 TFList2[i,4] <- V3
 TFList2[i,5] <- V4
 
     
TFList[i,6] <- log2((V1/(V1+V3))/(V2/(V2+V4)))
  TFList[i,7] <- -log2((V3/(V1+V3))/(V4/(V2+V4)))
  
  #TFList[i,8] <- log2((V1/(V1+V2))/(V3/(V3+V4)))
    #CTCF bound CTCF not bound 
#gain     V2`        V2 
#loss     V3        V4

#fd of enrichment of gain = (V1/V1+V3)/(V2/V2+V4)


 
}
   TFList$fdr_gain <- p.adjust(TFList[,4], "BH")
   TFList$fdr_loss <- p.adjust(TFList[,5], "BH")
   
gained_TF <- TFList %>% subset(fdr_gain < 0.05)
lost_TF <- TFList %>% subset(fdr_loss < 0.05)

gained_TF <- select(gained_TF, c("TF_name", "V2","V3","V6","fdr_gain"))
colnames(gained_TF) <- c("TF_Name", "gain_count","lost_count", "Enrichment","FDR")
gained_TF$group <- "gained"
lost_TF <- select(lost_TF, c("TF_name", "V2","V3","V7","fdr_loss"))
colnames(lost_TF) <- c("TF_Name", "gain_count","lost_count","Enrichment","FDR")
lost_TF$group <- "lost"
#TF_norm %>% subset( overlap_TF_gain > 0) %>% length

df_final <- rbind(gained_TF, lost_TF)

df_final <- df_final[order(-df_final$Enrichment),]
#df_final <- melt(df_final, id=c(1))


        df_final_gain_loss_bg <- df_final    
        df_final_gain_loss_bg
        
        temp <- TFList[c(1:4,6,8,5,7,9)]
        
        colnames(temp)[c(2:9)] <- c("gain_count", "loss_count", "P-value_gain","Enrichment_gain","fdr_gain", "P-value_loss", "Enrichment_loss","fdr_loss")
        
        colnames(temp)[c(2:9)] <- paste0(colnames(temp)[c(2:9)],"_diff_bg")
        
TFList_diff_bg  <- temp

colnames(TFList2)[c(2:dim(TFList2)[2])] <- c(paste0(rep("V", 4),1:4, "_gain"))
colnames(TFList2)[c(2:dim(TFList2)[2])] <- paste0(colnames(TFList2)[c(2:dim(TFList2)[2])],"_diff_bg")

TFList_diff_bg_merge <- merge(TFList_diff_bg,TFList2)

TFList_diff_bg_merge <- TFList_diff_bg_merge[c(1,4:dim(TFList_diff_bg_merge)[2])]

```


##Enrichment analysis for correct the distribution bias

            TF1 | sum(TF2 | TF3 .....)
    Gain
    Lost  

####get the TFList from above analysis

#####we need number of gain and lost for each TF

```{r}
#TFList <- aa
TFList <- select(TFList, c("TF_name","V2","V3"))

TFList2 <- TFList
for(i in 1:dim(TFList)[1]){
 
  
  
  successBG <- sum(TFList$V2)
  failBG <- sum(TFList$V3)
  overlap_TF_gain <- TFList[i,2]
  overlap_TF_loss <-TFList[i,3]
  
  
  TFList[i,4]<-phyper((overlap_TF_gain)-1,successBG,failBG,overlap_TF_gain+overlap_TF_loss, lower.tail = FALSE)
#  
  TFList[i,5]<-phyper((overlap_TF_loss)-1,failBG,successBG,overlap_TF_gain+overlap_TF_loss, lower.tail = FALSE)
   V1 = overlap_TF_gain
   V2 = successBG-overlap_TF_gain
   V3 = overlap_TF_loss
   V4 = failBG-overlap_TF_loss
      
 TFList[i,6] <- log2((V1/(V1+V3))/(V2/(V2+V4)))
   TFList[i,7] <- -log2((V3/(V1+V3))/(V4/(V2+V4)))
   
     TFList2[i,2] <- V1
 TFList2[i,3] <- V2
 TFList2[i,4] <- V3
 TFList2[i,5] <- V4
 
    #CTCF bound other TF bound except CTCF 
#gain     V2`        V2 
#loss     V3        V4

#fd of enrichment of gain = (V1/V1+V3)/(V2/V2+V4)


 
}


 TFList$fdr_gain <- p.adjust(TFList[,4], "BH")
   TFList$fdr_loss <- p.adjust(TFList[,5], "BH")
   
gained_TF <- TFList %>% subset(fdr_gain < 0.05)
lost_TF <- TFList %>% subset(fdr_loss < 0.05)

gained_TF <- select(gained_TF, c("TF_name", "V2","V3","V6","fdr_gain"))
colnames(gained_TF) <- c("TF_Name","gain_count","lost_count", "Enrichment","FDR")
gained_TF$group <- "gained"
lost_TF <- select(lost_TF, c("TF_name", "V2","V3","V7","fdr_loss"))
colnames(lost_TF) <- c("TF_Name", "gain_count","lost_count","Enrichment","FDR")
lost_TF$group <- "lost"
#TF_norm %>% subset( overlap_TF_gain > 0) %>% length

df_final <- rbind(gained_TF, lost_TF)

df_final <- df_final[order(-df_final$Enrichment),]
#df_final <- melt(df_final, id=c(1))


        df_final_gain_loss_bg_corr_bias <- df_final   
        df_final_gain_loss_bg_corr_bias
        
        
          temp <- TFList[c(1:4,6,8,5,7,9)]
        
        colnames(temp)[c(2:9)] <- c("gain_count", "loss_count", "P-value_gain","Enrichment_gain","fdr_gain", "P-value_loss", "Enrichment_loss","fdr_loss")
        
        colnames(temp)[c(2:9)] <- paste0(colnames(temp)[c(2:9)],"_diff_bg_bias_corrected")
        
TFList_diff_bg_bias_corrected  <- temp

TFList2

colnames(TFList2)[c(2:dim(TFList2)[2])] <-c(paste0(rep("V", 4),1:4, "_gain"))
colnames(TFList2)[c(2:dim(TFList2)[2])] <- paste0(colnames(TFList2)[c(2:dim(TFList2)[2])],"_diff_bg_bias_corrected")

TFList_diff_bg_bias_corrected_merge <- merge(TFList_diff_bg_bias_corrected,TFList2)

TFList_diff_bg_bias_corrected_merge <- TFList_diff_bg_bias_corrected_merge[c(1,4:dim(TFList_diff_bg_bias_corrected_merge)[2])]

TFList_diff_bg_bias_corrected_merge

```



#selecting gain and all others as bg 

```{r}

#load TF filenormalized to 400 bp from middle
##hg19 based
#TF_norm <- fread("TF.binding.sites.extended.range.200bp.from.middle.bed")

#colnames(TF_norm) <- c("chr","start","end","TF_name","score")

#TF_norm <- makeGRangesFromDataFrame(TF_norm, keep.extra.columns=TRUE)

predicted_element <- fread("enhancers/02.enhancer_vs_expression.table.csv")

predicted_element <- subset(predicted_element, !is.na("Peak"))
predicted_element <- unique(predicted_element)
element_width=1000



predicted_element$start <- predicted_element$start+ as.integer((predicted_element$end-predicted_element$start)/2) 
predicted_element$end <- predicted_element$start 
predicted_element$start <- predicted_element$start - (element_width/2) +1
predicted_element$end  <- predicted_element$end + (element_width/2)


predicted_element<- select(predicted_element, c("chr", "start", "end", "Peak","predicted_gain","predicted_loss"))
predicted_element <- subset(predicted_element, Peak!="")

gain <- subset(predicted_element, predicted_gain==T & predicted_loss==F)

lost <- subset(predicted_element, predicted_gain!=T )




gain <- makeGRangesFromDataFrame(gain, keep.extra.columns=TRUE)
lost <- makeGRangesFromDataFrame(lost, keep.extra.columns=TRUE)
	
TFList <- TF_norm %>% as.data.frame() %>%select(TF_name) %>% unique()

TFList2 <- TFList

for(i in 1:dim(TFList)[1]){
  TF_overlap <- subset(TF_norm, TF_name==TFList[i,1])
  overlap_TF_gain <- countOverlaps(  gain, TF_overlap, ignore.strand=TRUE,type="any")
    overlap_TF_gain <- as.data.frame(overlap_TF_gain)
 overlap_TF_gain <- subset(overlap_TF_gain, overlap_TF_gain > 0) %>% nrow()
 
overlap_TF_loss <- countOverlaps(lost , TF_overlap,  ignore.strand=TRUE,type="any")
  overlap_TF_loss <- as.data.frame(overlap_TF_loss)
 overlap_TF_loss <- subset(overlap_TF_loss, overlap_TF_loss > 0) %>% nrow() 
 
 TFList[i,2] <- overlap_TF_gain
 TFList[i,3] <- overlap_TF_loss
 
 TFList[i,4]<-phyper((overlap_TF_gain)-1,length(gain),length(lost),overlap_TF_gain+overlap_TF_loss, lower.tail = FALSE)
 
  V1 = overlap_TF_gain
  V2 = length(gain)-overlap_TF_gain
  V3 = overlap_TF_loss
  V4 = length(lost)-overlap_TF_loss
     
TFList[i,5] <- log2((V1/(V1+V3))/(V2/(V2+V4)))
 
 TFList2[i,2] <- V1
 TFList2[i,3] <- V2
 TFList2[i,4] <- V3
 TFList2[i,5] <- V4
  
 
}
   TFList$fdr_gain <- p.adjust(TFList[,4], "BH")

   
gained_TF <- TFList %>% subset(fdr_gain < 0.05)

gained_TF <- select(gained_TF, c("TF_name", "V2","V3","V5","fdr_gain"))
colnames(gained_TF) <- c("TF_Name","gain_count","not_gained_count", "Enrichment","FDR")
gained_TF$group <- "gained"


gained_TF_tot_bg <-gained_TF


gained_TF_tot_bg


temp <- TFList
        
        colnames(temp)[c(2:6)] <- c("gain_count", "not_gain_count", "P-value_gain","Enrichment_gain","fdr_gain")
        
        colnames(temp)[c(2:6)] <- paste0(colnames(temp)[c(2:6)],"_tot_bg")
        
TFList_gain_tot_bg  <- temp

TFList2


colnames(TFList2)[c(2:dim(TFList2)[2])] <- c(paste0(rep("V", 4),1:4))
colnames(TFList2)[c(2:dim(TFList2)[2])] <- paste0(colnames(TFList2)[c(2:dim(TFList2)[2])],"_tot_bg_gain")

TFList_gain_tot_bg_merge <- merge(TFList_gain_tot_bg,TFList2)

TFList_gain_tot_bg_merge <- TFList_gain_tot_bg_merge[c(1,4:dim(TFList_gain_tot_bg_merge)[2])]

TFList_gain_tot_bg_merge

```



## enrichment analysis for correct the distribution bias

                  TF1 | sum(TF2 | TF3 .....)
    Gain
    not gained  

####get the TFList from above analysis

#####we need number of gain and not gained for each TF

```{r}
#TFList <- aa
TFList <- select(TFList, c("TF_name","V2","V3"))

TFList2 <- TFList
for(i in 1:dim(TFList)[1]){
 
  
  
  successBG <- sum(TFList$V2)
  failBG <- sum(TFList$V3)
  overlap_TF_gain <- TFList[i,2]
  overlap_TF_loss <-TFList[i,3]
  
  
  TFList[i,4]<-phyper((overlap_TF_gain)-1,successBG,failBG,overlap_TF_gain+overlap_TF_loss, lower.tail = FALSE)
  
   V1 = overlap_TF_gain
   V2 = successBG-overlap_TF_gain
   V3 = overlap_TF_loss
   V4 = failBG-overlap_TF_loss
      
 TFList[i,5] <- log2((V1/(V1+V3))/(V2/(V2+V4)))
 
  TFList2[i,2] <- V1
 TFList2[i,3] <- V2
 TFList2[i,4] <- V3
 TFList2[i,5] <- V4
   
 
}


 TFList$fdr_gain <- p.adjust(TFList[,4], "BH")

   
gained_TF <- TFList %>% subset(fdr_gain < 0.05)


gained_TF <- select(gained_TF, c("TF_name", "V2","V3","V5","fdr_gain"))
colnames(gained_TF) <- c("TF_Name","gain_count","not_gained_count", "Enrichment","FDR")
gained_TF$group <- "gained"


        gained_TF_tot_bg_corr_bias <- gained_TF   
        
        gained_TF_tot_bg_corr_bias
        
        
        temp <- TFList
        
        colnames(temp)[c(2:6)] <- c("gain_count", "not_gain count", "P-value_gain","Enrichment_gain","fdr_gain")
        
        colnames(temp)[c(2:6)] <- paste0(colnames(temp)[c(2:6)],"_tot_bg_bias_corrected")
        
TFList_gain_tot_bg_bias_corrected  <- temp

TFList2

colnames(TFList2)[c(2:dim(TFList2)[2])] <- c(paste0(rep("V", 4),1:4))
colnames(TFList2)[c(2:dim(TFList2)[2])] <- paste0(colnames(TFList2)[c(2:dim(TFList2)[2])],"_tot_bg_bias_corrected_gain")

TFList_gain_tot_bg_bias_corrected_merge <- merge(TFList_gain_tot_bg_bias_corrected,TFList2)

TFList_gain_tot_bg_bias_corrected_merge <- TFList_gain_tot_bg_bias_corrected_merge[c(1,4:dim(TFList_gain_tot_bg_bias_corrected_merge)[2])]

TFList_gain_tot_bg_bias_corrected_merge


```



#selecting loss and all others as bg 
```{r}

#load TF filenormalized to 400 bp from middle
##hg19 based
#TF_norm <- fread("TF.binding.sites.extended.range.200bp.from.middle.bed")

#colnames(TF_norm) <- c("chr","start","end","TF_name","score")

#TF_norm <- makeGRangesFromDataFrame(TF_norm, keep.extra.columns=TRUE)

predicted_element <- fread("enhancers/02.enhancer_vs_expression.table.csv")

predicted_element <- unique(predicted_element)
element_width=1000



predicted_element$start <- predicted_element$start+ as.integer((predicted_element$end-predicted_element$start)/2) 
predicted_element$end <- predicted_element$start 
predicted_element$start <- predicted_element$start - (element_width/2) +1
predicted_element$end  <- predicted_element$end + (element_width/2)


predicted_element<- select(predicted_element, c("chr", "start", "end", "Peak","predicted_gain","predicted_loss"))
predicted_element <- subset(predicted_element, Peak!="")

gain <- subset(predicted_element, predicted_loss!=T )

lost <- subset(predicted_element, predicted_gain==F & predicted_loss==T)




gain <- makeGRangesFromDataFrame(gain, keep.extra.columns=TRUE)#not lost
lost <- makeGRangesFromDataFrame(lost, keep.extra.columns=TRUE) #lost
	
TFList <- TF_norm %>% as.data.frame() %>%select(TF_name) %>% unique()

TFList2 <- TFList

for(i in 1:dim(TFList)[1]){
  TF_overlap <- subset(TF_norm, TF_name==TFList[i,1])
  overlap_TF_gain <- countOverlaps(  gain, TF_overlap, ignore.strand=TRUE,type="any")
    overlap_TF_gain <- as.data.frame(overlap_TF_gain)
 overlap_TF_gain <- subset(overlap_TF_gain, overlap_TF_gain > 0) %>% nrow()
 
overlap_TF_loss <- countOverlaps(lost , TF_overlap,  ignore.strand=TRUE,type="any")
  overlap_TF_loss <- as.data.frame(overlap_TF_loss)
 overlap_TF_loss <- subset(overlap_TF_loss, overlap_TF_loss > 0) %>% nrow() 
 
 TFList[i,2] <- overlap_TF_gain
 TFList[i,3] <- overlap_TF_loss
 
 
 TFList[i,4]<-phyper((overlap_TF_loss)-1,length(lost),length(gain),overlap_TF_gain+overlap_TF_loss, lower.tail = FALSE)
 
  V1 = overlap_TF_gain
  V2 = length(gain)-overlap_TF_gain
  V3 = overlap_TF_loss
  V4 = length(lost)-overlap_TF_loss
     

  TFList[i,5] <- -log2((V3/(V1+V3))/(V4/(V2+V4)))
  
 TFList2[i,2] <- V3
 TFList2[i,3] <- V4
 TFList2[i,4] <- V1
 TFList2[i,5] <- V2
}
  
   TFList$fdr_lost<- p.adjust(TFList[,4], "BH")
   

   
lost_TF <- TFList %>% subset(fdr_lost < 0.05)

lost_TF <- select(lost_TF, c("TF_name", "V2","V3","V5","fdr_lost"))
colnames(lost_TF) <- c("TF_Name","lost_count","not_lost_count", "Enrichment","FDR")
lost_TF$group <- "lost"


lost_TF_tot_bg <-lost_TF


lost_TF_tot_bg

   temp <- TFList[c(1,3,2,4:6)]
        
        colnames(temp)[c(2:6)] <- c("loss_count", "not_loss_count", "P-value_loss","Enrichment_loss","fdr_loss")
        
        colnames(temp)[c(2:6)] <- paste0(colnames(temp)[c(2:6)],"_tot_bg")
        
TFList_loss_tot_bg <- temp

TFList2



colnames(TFList2)[c(2:dim(TFList2)[2])] <- c(paste0(rep("V", 4),1:4))
colnames(TFList2)[c(2:dim(TFList2)[2])] <- paste0(colnames(TFList2)[c(2:dim(TFList2)[2])],"_tot_bg_loss")

TFList_loss_tot_bg_merge <- merge(TFList_loss_tot_bg,TFList2)

TFList_loss_tot_bg_merge <- TFList_loss_tot_bg_merge[c(1,4:dim(TFList_loss_tot_bg_merge)[2])]

TFList_loss_tot_bg_merge


```

#merge total bg together
```{r}

lost_TF_tot_bg_temp <- lost_TF_tot_bg
gained_TF_tot_bg_temp <- gained_TF_tot_bg

colnames(gained_TF_tot_bg_temp)[c(2,3)] <- c("changed_count", "not_changed_count")
colnames(lost_TF_tot_bg_temp)[c(2,3)] <- c("changed_count", "not_changed_count")

final_tot_bg <- rbind(gained_TF_tot_bg_temp, lost_TF_tot_bg_temp)

final_tot_bg

```



## enrichment analysis for correct the distribution bias

            TF1 | sum(TF2 | TF3 .....)
    lost
    not lost  

####get the TFList from above analysis

#####we need number of lost and not lost for each TF

```{r}
#TFList <- aa
TFList <- select(TFList, c("TF_name","V2","V3"))
TFList2 <- TFList
for(i in 1:dim(TFList)[1]){
 
  
  
  successBG <- sum(TFList$V2)
  failBG <- sum(TFList$V3)
  overlap_TF_gain <- TFList[i,2]
  overlap_TF_loss <-TFList[i,3]
  
  
  TFList[i,4]<-phyper((overlap_TF_loss)-1,failBG,successBG,overlap_TF_gain+overlap_TF_loss, lower.tail = FALSE)
#  
 
   V1 = overlap_TF_gain
   V2 = successBG-overlap_TF_gain
   V3 = overlap_TF_loss
   V4 = failBG-overlap_TF_loss
      

   TFList[i,5] <- -log2((V3/(V1+V3))/(V4/(V2+V4)))

 TFList2[i,2] <- V3
 TFList2[i,3] <- V4
 TFList2[i,4] <- V1
 TFList2[i,5] <- V2
}


 
   TFList$fdr_loss <- p.adjust(TFList[,4], "BH")
   

lost_TF <- TFList %>% subset(fdr_loss < 0.05)


lost_TF <- select(lost_TF, c("TF_name", "V2","V3","V5","fdr_loss"))
colnames(lost_TF) <- c("TF_Name", "lost_count","not_lost_count","Enrichment","FDR")
lost_TF$group <- "lost"
#TF_norm %>% subset( overlap_TF_gain > 0) %>% length


lost_TF_tot_bg_corr_bias <- lost_TF   
        
        lost_TF_tot_bg_corr_bias  
        
        
         temp <- TFList[c(1,3,2,4:6)]
        
        colnames(temp)[c(2:6)] <- c("loss_count", "not_loss_count", "P-value_loss","Enrichment_loss","fdr_loss")
        
        colnames(temp)[c(2:6)] <- paste0(colnames(temp)[c(2:6)],"_tot_bg_bias_corrected")
        
TFList_loss_tot_bg_bias_corrected  <- temp

TFList2

colnames(TFList2)[c(2:dim(TFList2)[2])] <- c(paste0(rep("V", 4),1:4))
colnames(TFList2)[c(2:dim(TFList2)[2])] <- paste0(colnames(TFList2)[c(2:dim(TFList2)[2])],"_tot_bg_bias_corrected_loss")

TFList_loss_tot_bg_bias_corrected_merge <- merge(TFList_loss_tot_bg_bias_corrected,TFList2)

TFList_loss_tot_bg_bias_corrected_merge <- TFList_loss_tot_bg_bias_corrected_merge[c(1,4:dim(TFList_loss_tot_bg_bias_corrected_merge)[2])]

TFList_loss_tot_bg_bias_corrected_merge
```


```{r}

#merge total bg together


gained_TF_tot_bg_corr_bias_temp <- gained_TF_tot_bg_corr_bias
lost_TF_tot_bg_corr_bias_temp <- lost_TF_tot_bg_corr_bias

colnames(gained_TF_tot_bg_corr_bias_temp)[c(2,3)] <- c("changed_count", "not_changed_count")
colnames(lost_TF_tot_bg_corr_bias_temp)[c(2,3)] <- c("changed_count", "not_changed_count")

final_tot_bgcorr_bias <- rbind(gained_TF_tot_bg_corr_bias_temp, lost_TF_tot_bg_corr_bias_temp)


final_tot_bgcorr_bias

```


####merge all the four test results

```{r}
#df_final_gain_loss_bg
#df_final_gain_loss_bg_corr_bias
#final_tot_bgcorr_bias
#final_tot_bg

#meke all=T kiyala thibbe man mekuwa mathaka ne eka mokada thiyenne kiyala
#eka thibbama only fdr filtered ewa witharak neme enne kiyala mata passe hithunnu nisa man eka makuwa
#eththatama eka mulin demme ey mathaka ne den

final_TF_all_4_test <- merge(df_final_gain_loss_bg, df_final_gain_loss_bg_corr_bias, by="TF_Name", suffixes=c("_diff_bg","_diff_bg_bias_corrected"))

final_TF_all_4_test_temp <- merge(final_tot_bg,final_tot_bgcorr_bias, by="TF_Name", suffixes=c("_tot_bg","_tot_bg_bias_corrected"))


final_TF_all_4_test <- merge(final_TF_all_4_test, final_TF_all_4_test_temp, by="TF_Name")

final_TF_all_4_test

write.csv(final_TF_all_4_test, file="Enhancer_TF_analysis_all4_tests.csv")
```


####Create table without filtering by FDR
```{r}


temp <- cbind(TFList_gain_tot_bg, TFList_loss_tot_bg)
temp <- temp[,-7]
temp2 <- cbind(TFList_gain_tot_bg_bias_corrected, TFList_loss_tot_bg_bias_corrected)
temp2 <- temp2[,-7]
TFList_diff_bg_bias_corrected

final_tf_df <- merge(TFList_diff_bg,TFList_diff_bg_bias_corrected)

final_tf_df <- merge(final_tf_df, temp)

final_tf_df <- merge(final_tf_df, temp2)

write.csv(final_tf_df, file="Enhancer_TF_analysis_all4_tests_without_filtering_by_FDR.csv")
```


####Create table without filtering by FDR with all counts in contingenecy table
```{r}

temp <- cbind(TFList_gain_tot_bg_merge, TFList_loss_tot_bg_merge)
temp <- temp[,-9]
temp2 <- cbind(TFList_gain_tot_bg_bias_corrected_merge, TFList_loss_tot_bg_bias_corrected_merge)
temp2 <- temp2[,-9]


final_tf_df <- merge(TFList_diff_bg_merge,TFList_diff_bg_bias_corrected_merge)

final_tf_df <- merge(final_tf_df, temp)

final_tf_df <- merge(final_tf_df, temp2)

write.csv(final_tf_df, file="Enhancer_TF_analysis_all4_tests_without_filtering_by_FDR_add_cont_table.csv")



```

###Fgiure 5A
####Create heatmap for TF enrichment
```{r, fig.height=10, fig.width=5}


df_tf <- read.csv( file="Enhancer_TF_analysis_all4_tests.csv")

df_tf$X <- NULL
df_tf <- subset(df_tf, !is.na(group_tot_bg))

df_tf <- df_tf[order(-df_tf$Enrichment_diff_bg),]

df_tf_enrichment <- select(df_tf, c(1,4,9,14,19))

df_tf_enrichment <- melt(df_tf_enrichment, id=c(1))

df_tf_fdr <- select(df_tf, c(1,5,10,15,20))

df_tf_fdr<- melt(df_tf_fdr, id=c(1))

d.matrix <- data.frame(df_tf_enrichment,df_tf_fdr)

d.matrix <- d.matrix[c(1:3,6)]

colnames(d.matrix) <- c("TF_Name", "group", "Enrichment", "fdr")

d.matrix$Enrichment[is.na(d.matrix$Enrichment)] <- 0

#convert negative sign back to positive # enrichment 
d.matrix$Enrichment[d.matrix$Enrichment < 0] <- d.matrix$Enrichment[d.matrix$Enrichment < 0] * -1


length_TF <- d.matrix %>% select(TF_Name) %>% unique() 

figure_5b_tf_order <- length_TF
#pdf(file="C:/Users/pubud/Dropbox/thesis/graphs/fig 5/fig5_a_TF_enhancers.pdf", height=10,width=4) 
 ggplot(d.matrix, size = d.matrix$Enrichment, aes(x = d.matrix$group, size = d.matrix$Enrichment ,
 y= factor(d.matrix$TF_Name , levels = rev(as.character(length_TF$TF_Name))))) +
  geom_point(aes(colour =squish(log10(d.matrix$fdr),c(-20,0)))) +
  scale_colour_gradient("log10(FDR of\nenrichment)", low = "red",  high="yellow") +
  scale_size("log2 FD\nof enrichment",range = c(0, 10), breaks =c(seq(0,1.5,by = 0.5)))+
  guides(size = guide_legend(override.aes = list(colour = "black")))+
 # guides(size = guide_legend(override.aes = list(size=c(0,0.5))))+
 # theme_bw()+
   theme_minimal()+
  theme(legend.key.size = unit(1.5, 'lines'), text = element_text(size=15), axis.text.x=element_text(colour="white"),axis.text = element_text( size=14),axis.title.x=element_blank(),axis.title.y=element_blank(),legend.position="left", 
						panel.background = element_blank(), axis.line = element_line(colour = "black"),axis.ticks.length=unit(.25, "cm"), axis.ticks.x=element_blank())

#dev.off()
#getwd()

```


#Pathway enrichment analysis

##Figure 5B
  TF target genes analysis
```{r}

#load TF file normalized to 400 bp from middle
##hg19 based
TF_norm <- fread("TF.binding.sites.extended.range.200bp.from.middle.bed")

TF_list <- fread("Enhancer_TF_analysis_all4_tests.csv")
colnames(TF_norm) <- c("chr","start","end","TF_name","score")

TF_norm <- makeGRangesFromDataFrame(TF_norm, keep.extra.columns=TRUE)

predicted_element <- fread("02.enhancer_vs_expression.table.csv")

predicted_element <- unique(predicted_element)

predicted_element_ori <- predicted_element
element_width=1000



predicted_element$start <- predicted_element$start+ as.integer((predicted_element$end-predicted_element$start)/2) 
predicted_element$end <- predicted_element$start 
predicted_element$start <- predicted_element$start - (element_width/2) +1
predicted_element$end  <- predicted_element$end + (element_width/2)


predicted_element<- dplyr::select(predicted_element, c("chr", "start", "end", "Peak","predicted_gain","predicted_loss"))
predicted_element <- subset(predicted_element, Peak!="")

gain <- subset(predicted_element, predicted_gain==T & predicted_loss==F)

lost <- subset(predicted_element, predicted_gain==F & predicted_loss==T)




gain <- makeGRangesFromDataFrame(gain, keep.extra.columns=TRUE)
lost <- makeGRangesFromDataFrame(lost, keep.extra.columns=TRUE)
	
#remove comment if u need lost
#gain <- lost


 
  overlap_TF_gain <- findOverlaps(  gain, TF_norm, ignore.strand=TRUE,type="any")
  
queryhits <-  gain[ queryHits(overlap_TF_gain)] %>% as.data.frame()
 
 subhits <-  TF_norm[ subjectHits(overlap_TF_gain)] %>% as.data.frame()
 
 colnames(queryhits) <- paste0(c(colnames(queryhits)),"_query")
 
 
 df_gain_merge <- cbind(subhits ,queryhits)
 
 temp_gain_merge <- df_gain_merge %>% select(c("TF_name", "Peak_query"))
 
 #chnage gain lost below
 
 TF_list_gain <- subset(TF_list, TF_list$significant_all4_tests==T & TF_list$significant_all4_tests_type=="Gain")
 
 TF_list_gain <- select(TF_list_gain, c("Gene name", "Gene"))

    
```
 
 
 TF target genes analysis continues....
```{r}
#load file

predicted_element_ori_temp <- select(predicted_element_ori, c("Peak", "Gene.name"))

TF_gain_list_enh <- merge(merge(TF_list_gain, temp_gain_merge, by.x="Gene name",by.y="TF_name"), predicted_element_ori_temp, by.x="Peak_query", by.y="Peak")

TF_gain_list_enh_temp <- select(TF_gain_list_enh, c("Gene name", "Gene.name")) %>% unique()


colnames(TF_gain_list_enh_temp) <- c("TF_Name","Gene")


t.group <- data.frame(t(TF_gain_list_enh_temp))

#replace column names by first row values
colnames(t.group) <- as.character(unlist(t.group[1,]))

#remove first row
t.group <- t.group[-1, ]


#df[,names(df) == 'a'] will select all columns with name a
#unlist will convert above columns into 1 single vector
#unname will remove some stray rownames given to these vectors.
#unique(names(df)) will give you unique column names in df
#sapply will apply the inline function to all values of unique(names(df))

#Create a list with genes in the rows and TF in the columns( each column represent an TF and rows represent genes overlapped in that TF


t.group <- sapply(unique(names(t.group)), function(x) unname(unlist(t.group[,names(t.group)==x])))

#order list named (grp1 to grp5)
t.group <- t.group[order(names(t.group))]
colnames_t_grp <- names(t.group)
#list to data frame
t.group <-as.data.frame((stri_list2matrix(t.group)))



colnames(t.group) <- colnames_t_grp

tot_gain_tf_genes <- TF_gain_list_enh_temp %>% select(Gene) %>% unique()


write.table(t(t.group), file="gained_enhancer_TF_gsea.gmt", col.names = F, row.names = T, quote = F, sep = "\t")

write.table(tot_gain_tf_genes, file="gained_enh_TF/gained_enhancer_TF_genes.txt", col.names = F, row.names = F, quote = F, sep = "\t")

write.table(t.group, file="gained_enh_TF/gained_enhancer_TF_genes_all.txt", col.names = T, row.names = F, quote = F, sep = "\t")
###for lost 

gain <- lost


 
  overlap_TF_gain <- findOverlaps(  gain, TF_norm, ignore.strand=TRUE,type="any")
  
queryhits <-  gain[ queryHits(overlap_TF_gain)] %>% as.data.frame()
 
 subhits <-  TF_norm[ subjectHits(overlap_TF_gain)] %>% as.data.frame()
 
 colnames(queryhits) <- paste0(c(colnames(queryhits)),"_query")
 
 
 df_gain_merge <- cbind(subhits ,queryhits)
 
 temp_gain_merge <- df_gain_merge %>% select(c("TF_name", "Peak_query"))
 
 #chnage gain lost below
 
 TF_list_gain <- subset(TF_list, TF_list$significant_all4_tests==T & TF_list$significant_all4_tests_type=="Gain")
 
 TF_list_gain <- select(TF_list_gain, c("Gene name", "Gene"))
 
 
 #load file

predicted_element_ori_temp <- select(predicted_element_ori, c("Peak", "Gene.name"))

TF_gain_list_enh <- merge(merge(TF_list_gain, temp_gain_merge, by.x="Gene name",by.y="TF_name"), predicted_element_ori_temp, by.x="Peak_query", by.y="Peak")

TF_gain_list_enh_temp <- select(TF_gain_list_enh, c("Gene name", "Gene.name")) %>% unique()


colnames(TF_gain_list_enh_temp) <- c("TF_Name","Gene")


t.group <- data.frame(t(TF_gain_list_enh_temp))

#replace column names by first row values
colnames(t.group) <- as.character(unlist(t.group[1,]))

#remove first row
t.group <- t.group[-1, ]


#df[,names(df) == 'a'] will select all columns with name a
#unlist will convert above columns into 1 single vector
#unname will remove some stray rownames given to these vectors.
#unique(names(df)) will give you unique column names in df
#sapply will apply the inline function to all values of unique(names(df))

#Create a list with genes in the rows and TF in the columns( each column represent an TF and rows represent genes overlapped in that TF


t.group <- sapply(unique(names(t.group)), function(x) unname(unlist(t.group[,names(t.group)==x])))

#order list named (grp1 to grp5)
t.group <- t.group[order(names(t.group))]
colnames_t_grp <- names(t.group)
#list to data frame
t.group <-as.data.frame((stri_list2matrix(t.group)))



colnames(t.group) <- colnames_t_grp

tot_gain_tf_genes <- TF_gain_list_enh_temp %>% select(Gene) %>% unique()


write.table(t(t.group), file="lost_enhancer_TF_gsea.gmt", col.names = F, row.names = T, quote = F, sep = "\t")

write.table(tot_gain_tf_genes, file="lost_enh_TF/lost_enhancer_TF_genes.txt", col.names = F, row.names = F, quote = F, sep = "\t")

write.table(t.group, file="lost_enh_TF/lost_enhancer_TF_genes_all.txt", col.names = T, row.names = F, quote = F, sep = "\t")
TF_gain_list_enh_unique <- TF_gain_list_enh_temp

```


 
 
 Select only up regulated target genes of Tfs that are enriched in gained enhancers for cpdb analysis
 
```{r}

gain_target_list <- fread( file="gained_enh_TF/gained_enhancer_TF_genes.txt", header = F)

####RNA seq data load
	vhlev_H <- read.csv(file="RNA.seq.hypoxia.VHLandwildtype.control.is.EV.csv")
	vhlev_H$rn <- gsub("\\..*","",vhlev_H$rn)
	vhlev_H <- select(vhlev_H, c("rn","log2FoldChange","padj"))
	colnames(vhlev_H) <- c("rn","log2FoldChange_vhlev_H","padj_vhlev_H")

	
	vhlev_N <- read.csv(file="RNAseqRCC4/RNA.seq.normoxia.VHLandwildtype.control.is.EV.csv")
	vhlev_N$rn <- gsub("\\..*","",vhlev_N$rn)
	vhlev_N <- select(vhlev_N, c("rn","log2FoldChange","padj"))
	colnames(vhlev_N) <- c("rn","log2FoldChange_vhlev_N","padj_vhlev_N")
	
	patients_diff_genes <- read.csv(file="ordered down regulated genes paired conventional patients.csv")
	patients_diff_genes <- select(patients_diff_genes, c("X","log2FoldChange","padj"))
	colnames(patients_diff_genes) <- c("rn","log2FoldChange_patients","padj_patients")
	
	gene.names <- fread("gene_reference.txt", header = F)
gene.names$V1 <- word(gene.names$V1,1, sep = "\\.")

patients_diff_genes <- merge(patients_diff_genes, gene.names, by.x="rn", by.y="V1")

patients_diff_genes <- subset(patients_diff_genes, padj_patients < 0.05)


gain_target_list <- merge(gain_target_list, patients_diff_genes, by.x="V1", by.y="V3")

gain_target_list <- gain_target_list %>% subset(log2FoldChange_patients > 0)


write.table(gain_target_list$V1, file="gained_enh_TF/gained_enhancer_TF_target_upregulated_genes_cpdb.txt", col.names = F, row.names = F, quote = F, sep = "\t")

```
 
 
 
 Select only down regulated target genes of Tfs that are enriched in lost enhancers for cpdb analysis
 
```{r}

loss_target_list <- fread( file="enhancers/lost_enh_TF/lost_enhancer_TF_genes.txt", header = F)


loss_target_list <- merge(loss_target_list, patients_diff_genes, by.x="V1", by.y="V3")

loss_target_list <- loss_target_list %>% subset(log2FoldChange_patients < 0)


write.table(loss_target_list$V1, file="lost_enh_TF/lost_enhancer_TF_target_downregulated_genes_cpdb.txt", col.names = F, row.names = F, quote = F, sep = "\t")

```

####Create bg list for enhancer cpdb enrichment analysis

#bg 1 is the default bg in cpdb
#bg 2 all significant genes

background 2
```{r}


write.table(patients_diff_genes$V3, file="gained_enh_TF/bg2_all_significant_genes_fdr_0.05_cpdb.txt", col.names = F, row.names = F, quote = F, sep = "\t")
```


bg 3 all  genes associated with enhancers
```{r}
enhancer_genes <- fread("enhancers/02.enhancer_vs_expression.table.csv")

enhancer_genes <- subset(enhancer_genes, Peak!="")
enhancer_genes <- select(enhancer_genes, Gene.name) %>% unique
.

write.table(enhancer_genes, file="enhancers/gained_enh_TF/bg3_all_enhancer_genes_cpdb.txt", col.names = F, row.names = F, quote = F, sep = "\t")
```


find pathways enriched in 3 bgs

gained up

```{r}

#Do the CPDB pathway analysis using these backgrounds and up regulated genes
#download the .TAB file from the server and rename them accordingly
bg1 <- fread("gained_enh_TF/up_regulated_gained_with_bg1.tab")
bg1 <- select(bg1, c("pathway","external_id"))
bg2 <- fread("gained_enh_TF/up_regulated_gained_with_bg2.tab")
bg2 <- select(bg2, c("pathway","external_id"))
bg3 <- fread("gained_enh_TF/up_regulated_gained_with_bg3.tab")
bg3 <- select(bg3, c("pathway","external_id"))

df_merge_pathways <- merge(merge(bg1,bg2, by="external_id"), bg3, by="external_id")

write.table(df_merge_pathways$pathway, file="enhancers/gained_enh_TF/common_pathwyas_3bgs_gained_upregulated_genes_cpdb.txt", col.names = F, row.names = F, quote = F, sep = "\t")

```

```{r}
library(matrixStats)
```



find pathways enriched in 3 backgrounds for sup table 4
```{r}
bg1 <- fread("gained_enh_TF/up_regulated_gained_with_bg1.tab")
bg1 <- select(bg1, c("pathway","external_id","q-value","source","members_input_overlap"))
bg2 <- fread("gained_enh_TF/up_regulated_gained_with_bg2.tab")
bg2 <- select(bg2, c("external_id","q-value"))
colnames(bg2) <-  c("external_id","q-value2")
bg3 <- fread("gained_enh_TF/up_regulated_gained_with_bg3.tab")
bg3 <- select(bg3, c("external_id","q-value"))
colnames(bg3) <-  c("external_id","q-value3")

df_merge_pathways <- merge(merge(bg1,bg2, by="external_id"), bg3, by="external_id")

rowmax_mat <- as.matrix(select(df_merge_pathways, c("q-value","q-value2", "q-value3")))
df_merge_pathways$max_fdr <- rowMaxs(rowmax_mat)

df_merge_pathways <- subset(df_merge_pathways, max_fdr < 0.05)

df_merge_pathways <- df_merge_pathways[order(df_merge_pathways$max_fdr)]

df_merge_pathways <- df_merge_pathways[,c(2,8,4:5)]
write.table(df_merge_pathways, file="tables/common_pathwyas_3bgs_gained_upregulated_genes_cpdb.txt", col.names = T, row.names = F, quote = F, sep = "\t")
```

create bar plor for figure
```{r, fig.width=10}

bg1 <- fread("gained_enh_TF/up_regulated_gained_with_bg1.tab")
#bg1 <- select(bg1, c("pathway","external_id"))
bg2 <- fread("gained_enh_TF/up_regulated_gained_with_bg2.tab")
#bg2 <- select(bg2, c("pathway","external_id"))
bg3 <- fread("gained_enh_TF/up_regulated_gained_with_bg3.tab")
#bg3 <- select(bg3, c("pathway","external_id"))

df_merge_pathways <- merge(merge(bg1,bg2, by="external_id",  suffixes=c("_bg1","_bg2")), bg3, by="external_id")

df_merge_pathways <- select(df_merge_pathways, c("pathway", "q-value_bg1","q-value_bg2","q-value"))

df_merge_pathways$max_FDR <- rowMins(as.matrix(df_merge_pathways[,c(2:4)]))

df_merge_pathways <- df_merge_pathways[order(df_merge_pathways$max_FDR)]

df_merge_pathways <- head(df_merge_pathways, n=20)


pdf(file="fig5_c_TF_enhancers_up_target_genes_cpdb.pdf", height=12,width=15) 
ggplot(df_merge_pathways, aes(x = factor(df_merge_pathways$pathway, levels = as.character(rev(df_merge_pathways$pathway))), y = log10(df_merge_pathways$max_FDR))) +
   
  #guides(fill=guide_legend(title="log P Value"))+
  geom_bar(stat = "identity", width = 0.5, colour="black", fill="black")+
 
  theme_bw()+
  # scale_fill_manual(values=c( "blue"))+
  scale_fill_gradient(low="#800033",high="#ff80b3", name = "FDR")+
    theme(legend.key.size = unit(1.5, 'lines'), axis.text.x = element_text( size=16),text = element_text(size=20), legend.position="right",panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),axis.ticks.length=unit(.25, "cm"))+
  coord_flip()+
  scale_y_reverse()
dev.off()


 


```

lost down

```{r}

bg1 <- fread("enhancers/lost_enh_TF/down_regulated_lost_with_bg1.tab")
bg1 <- select(bg1, c("pathway","external_id"))
bg2 <- fread("enhancers/lost_enh_TF/down_regulated_lost_with_bg2.tab")
bg2 <- select(bg2, c("pathway","external_id"))
bg3 <- fread("enhancers/lost_enh_TF/down_regulated_lost_with_bg3.tab")
bg3 <- select(bg3, c("pathway","external_id"))

df_merge_pathways <- merge(merge(bg1,bg2, by="external_id"), bg3, by="external_id")

write.table(df_merge_pathways$pathway, file="enhancers/lost_enh_TF/common_pathwyas_3bgs_lost_downregulated_genes_cpdb.txt", col.names = F, row.names = F, quote = F, sep = "\t")

```

```{r, fig.width=15}

bg1 <- fread("enhancers/lost_enh_TF/down_regulated_lost_with_bg1.tab")
#bg1 <- select(bg1, c("pathway","external_id"))
bg2 <- fread("enhancers/lost_enh_TF/down_regulated_lost_with_bg2.tab")
#bg2 <- select(bg2, c("pathway","external_id"))
bg3 <- fread("enhancers/lost_enh_TF/down_regulated_lost_with_bg3.tab")
#bg3 <- select(bg3, c("pathway","external_id"))

df_merge_pathways <- merge(merge(bg1,bg2, by="external_id",  suffixes=c("_bg1","_bg2")), bg3, by="external_id")

df_merge_pathways <- select(df_merge_pathways, c("pathway", "q-value_bg1","q-value_bg2","q-value"))

df_merge_pathways$max_FDR <- rowMins(as.matrix(df_merge_pathways[,c(2:4)]))

df_merge_pathways <- df_merge_pathways[order(df_merge_pathways$max_FDR)]

df_merge_pathways <- head(df_merge_pathways, n=20)

df_merge_pathways <- subset(df_merge_pathways, df_merge_pathways$max_FDR < 0.05)

#df_merge_pathways$pathway = str_wrap(df_merge_pathways$pathway, width = 35)
pdf(file="C:/Users/pubud/Dropbox/thesis/graphs/fig 5/fig5_d_TF_enhancers_down_target_genes_cpdb.pdf", height=3,width=12) 
ggplot(df_merge_pathways, aes(x = factor(df_merge_pathways$pathway, levels = as.character(rev(df_merge_pathways$pathway))), y = log10(df_merge_pathways$max_FDR))) +
   
  #guides(fill=guide_legend(title="log P Value"))+
  geom_bar(stat = "identity", width = 0.5, colour="black", fill="black")+
 
  theme_bw()+
  # scale_fill_manual(values=c( "blue"))+
  scale_fill_gradient(low="#800033",high="#ff80b3", name = "FDR")+
    theme(legend.key.size = unit(1.5, 'lines'), axis.text.x = element_text( size=16),text = element_text(size=20), legend.position="right",panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),axis.ticks.length=unit(.25, "cm"))+
  coord_flip()+
  scale_y_reverse()
dev.off()


 


```


####create rnk file for GSEA analysis 
```{r}

library(IHW)
gene.names <- fread("gene_reference.txt", header = F)
gene.names$V1 <- word(gene.names$V1,1, sep = "\\.")

patients.diff <- read.csv("ordered down regulated genes paired conventional vhl mutated patients IHW weights.csv")
#pat file name new(diddnt do any change to the code but result is diff than previous)
#ordered down regulated genes paired conventional vhl mutated patients_2018.12.17.csv

#get max of IHW
#remove everything less than 10% of IHW max
IHWcutoff <- max(patients.diff$weight) * 10 /100

resIHW_final <- subset(patients.diff, weight >= IHWcutoff)

gene.names_temp <- subset(gene.names, V2=="protein_coding")

#patients.diff <- subset(patients.diff, padj < 0.1)

diff_gene_rnk_df <- merge(resIHW_final,gene.names_temp, by.x="X", by.y="V1")

diff_gene_rnk_df_filt <- diff_gene_rnk_df

diff_gene_rnk_df <- select(diff_gene_rnk_df, c(V3, log2FoldChange))

diff_gene_rnk_df <- diff_gene_rnk_df[order(-diff_gene_rnk_df$log2FoldChange),]


write.table(diff_gene_rnk_df, file="conventional_patients_diff_genes_fdr_0.1_for_gsea.rnk", col.names = F, row.names = F, quote = F, sep = "\t")

```



###Perform the GSEA analysis with above rank file and .gmt files created by pervious steps
Enhancer TF target genes GSEa analysis bar graph

###Figure 5A bar plot

```{r, fig.height=8, fig.width=6}
gain_up <- fread("gained_enh_TF/gained_enh_TF.patients_.GseaPreranked.1552793757961/gsea_report_for_na_pos_1552793757961.xls")

loss_down <- fread("lost_enh_TF/lost_enh_TF.Gsea_patients_Preranked.1551210549785/gsea_report_for_na_neg_1551210549785.xls")

df_rbind <- rbind(gain_up, loss_down)

write.csv(df_rbind, file="patients_tf_target_genes_gsea_merge.csv", row.names = F)


pdf(file="fig5_b_TF_enhancers_gsea_order_by_tfs.pdf", height=6,width=4) 
ggplot(df_rbind, aes(x = factor(df_rbind$NAME, levels = as.character(rev(figure_5b_tf_order$TF_Name))), y = df_rbind$NES, fill=(df_rbind$"FDR q-val"))) +
  #guides(fill=guide_legend(title="log P Value"))+
  geom_bar(stat = "identity", width = 0.5)+
  theme_bw()+
  scale_fill_gradient(low="#800033",high="#ff80b3", name = "FDR")+
    theme(legend.key.size = unit(1.5, 'lines'), axis.text.x = element_text( size=14),text = element_text(size=16), axis.ticks.y=element_blank(),axis.line.y=element_blank(),legend.position="right",panel.grid.major = element_blank(), panel.grid.minor = element_blank())+
  coord_flip()
dev.off()

```

